File: archive/code/src/harness/providers/codex/actors.ts
import type {
  HarnessEvent,
  RunActor,
} from "../../events.js";
import {
  asRecord,
  stringArrayField,
  stringField,
} from "./fields.js";
import type { CodexRunState } from "./types.js";

export function actorForCodexThread(
  state: CodexRunState,
  threadId: string | undefined,
  rootThreadId: string | undefined,
): RunActor {
  if (threadId === undefined) {
    return {
      threadId: null,
      role: "unknown",
      confidence: "unknown",
      label: "Unknown actor",
    };
  }
  if (rootThreadId === undefined && state.rootThreadId === undefined) {
    state.rootThreadId = threadId;
  }
  const effectiveRootThreadId = rootThreadId ?? state.rootThreadId;
  if (effectiveRootThreadId !== undefined && threadId === effectiveRootThreadId) {
    return {
      threadId,
      role: "root",
      confidence: "provider",
      label: "Main",
    };
  }
  const parentThreadId = state.agentParents?.[threadId] ?? effectiveRootThreadId ?? null;
  return {
    threadId,
    role: "helper",
    parentThreadId,
    confidence: "provider",
    label: state.agentLabels?.[threadId] ?? helperLabel(state, threadId),
  };
}

export function codexLifecycleEvents(
  state: CodexRunState,
  item: Record<string, unknown>,
  actor: RunActor,
  phase: "started" | "completed",
): HarnessEvent[] {
  if (stringField(item, "type") !== "collabAgentToolCall") return [];
  const tool = stringField(item, "tool");
  const senderThreadId = stringField(item, "senderThreadId") ?? actor.threadId ?? null;
  const receiverThreadIds = stringArrayField(item, "receiverThreadIds");

  if (tool === "spawnAgent" && phase === "completed") {
    const events: HarnessEvent[] = [];
    for (const childThreadId of receiverThreadIds) {
      registerAgent(state, childThreadId, senderThreadId, stringField(item, "model"));
      events.push({
        type: "agent_spawned",
        parentThreadId: senderThreadId ?? "",
        childThreadId,
        prompt: stringField(item, "prompt") ?? "",
        model: stringField(item, "model"),
        reasoningEffort: stringField(item, "reasoningEffort"),
        actor,
      });
    }
    return events;
  }

  if (tool === "wait" && phase === "started") {
    return [
      {
        type: "agent_wait_started",
        parentThreadId: senderThreadId ?? "",
        childThreadIds: receiverThreadIds,
        actor,
      },
    ];
  }

  if (tool === "wait" && phase === "completed") {
    return completedAgentEventsFromWait(state, item);
  }

  return [];
}

function completedAgentEventsFromWait(
  state: CodexRunState,
  item: Record<string, unknown>,
): HarnessEvent[] {
  const agentsStates = asRecord(item.agentsStates);
  const events: HarnessEvent[] = [];
  for (const [threadId, rawState] of Object.entries(agentsStates)) {
    const agentState = asRecord(rawState);
    if (stringField(agentState, "status") !== "completed") continue;
    if (state.completedAgents?.[threadId] === true) continue;
    const message = stringField(agentState, "message");
    if (message === undefined) continue;
    markAgentCompleted(state, threadId);
    events.push({
      type: "agent_completed",
      threadId,
      parentThreadId: state.agentParents?.[threadId] ?? null,
      result: message,
      actor: actorForCodexThread(state, threadId, state.rootThreadId),
    });
  }
  return events;
}

function registerAgent(
  state: CodexRunState,
  childThreadId: string,
  parentThreadId: string | null,
  _model: string | undefined,
): void {
  state.agentParents = state.agentParents ?? {};
  state.agentLabels = state.agentLabels ?? {};
  state.agentParents[childThreadId] = parentThreadId;
  state.agentLabels[childThreadId] = helperLabel(state, childThreadId);
}

export function markAgentCompleted(state: CodexRunState, threadId: string): void {
  state.completedAgents = state.completedAgents ?? {};
  state.completedAgents[threadId] = true;
}

function helperLabel(state: CodexRunState, threadId: string): string {
  state.agentLabels = state.agentLabels ?? {};
  const existing = state.agentLabels[threadId];
  if (existing !== undefined) return existing;
  const count = Object.keys(state.agentLabels).length + 1;
  const label = `Helper ${count}`;
  state.agentLabels[threadId] = label;
  return label;
}

File: src/codealmanac/integrations/harnesses/codex/actors.py
from codealmanac.engine.harnesses.models import (
    HarnessActorConfidence,
    HarnessActorRole,
    HarnessRunActor,
)
from codealmanac.integrations.harnesses.codex.state import CodexRunState


def actor_for_codex_thread(
    state: CodexRunState,
    thread_id: str | None,
) -> HarnessRunActor:
    if thread_id is None:
        return HarnessRunActor(
            thread_id=None,
            role=HarnessActorRole.UNKNOWN,
            confidence=HarnessActorConfidence.UNKNOWN,
            label="Unknown actor",
        )
    if state.root_thread_id is None:
        state.root_thread_id = thread_id
    if thread_id == state.root_thread_id:
        return HarnessRunActor(
            thread_id=thread_id,
            role=HarnessActorRole.ROOT,
            confidence=HarnessActorConfidence.PROVIDER,
            label="Main",
        )
    return HarnessRunActor(
        thread_id=thread_id,
        role=HarnessActorRole.HELPER,
        parent_thread_id=state.agent_parents.get(thread_id) or state.root_thread_id,
        confidence=HarnessActorConfidence.PROVIDER,
        label=helper_label(state, thread_id),
    )


def helper_label(state: CodexRunState, thread_id: str) -> str:
    existing = state.agent_labels.get(thread_id)
    if existing is not None:
        return existing
    label = f"Helper {len(state.agent_labels) + 1}"
    state.agent_labels[thread_id] = label
    return label
